crates/uv-interpreter/src/discovery.rs

```diff
@@ -0,0 +1,1113 @@
+use itertools::Itertools;
+use thiserror::Error;
+use tracing::{debug, instrument, trace};
+use uv_cache::Cache;
+use uv_fs::Simplified;
+use uv_warnings::warn_user_once;
+use which::which;
+
+use crate::implementation::ImplementationName;
+use crate::managed::toolchains_for_current_platform;
+use crate::py_launcher::py_list_paths;
+use crate::virtualenv::virtualenv_python_executable;
+use crate::virtualenv::{virtualenv_from_env, virtualenv_from_working_dir};
+use crate::{Interpreter, PythonVersion};
+use std::borrow::Cow;
+
+use std::collections::HashSet;
+use std::fmt::{self, Formatter};
+use std::num::ParseIntError;
+use std::{env, io};
+use std::{path::Path, path::PathBuf, str::FromStr};
+
+/// A request to find a Python interpreter.
+///
+/// See [`InterpreterRequest::from_str`].
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum InterpreterRequest {
+    /// A Python version without an implementation name e.g. `3.10`
+    Version(VersionRequest),
+    /// A path to a directory containing a Python installation, e.g. `.venv`
+    Directory(PathBuf),
+    /// A path to a Python executable e.g. `~/bin/python`
+    File(PathBuf),
+    /// The name of a Python executable (i.e. for lookup in the PATH) e.g. `foopython3`
+    ExecutableName(String),
+    /// A Python implementation without a version e.g. `pypy`
+    Implementation(ImplementationName),
+    /// A Python implementation name and version e.g. `pypy3.8`
+    ImplementationVersion(ImplementationName, VersionRequest),
+}
+
+/// The sources to consider when finding a Python interpreter.
+#[derive(Debug, Clone, PartialEq, Eq, Default)]
+pub enum SourceSelector {
+    #[default]
+    All,
+    Some(HashSet<InterpreterSource>),
+}
+
+/// A Python interpreter version request.
+#[derive(Copy, Clone, Debug, Default, PartialEq, Eq)]
+pub enum VersionRequest {
+    #[default]
+    Default,
+    Major(u8),
+    MajorMinor(u8, u8),
+    MajorMinorPatch(u8, u8, u8),
+}
+
+/// The policy for discovery of "system" Python interpreters.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
+pub enum SystemPython {
+    // Do not allow a system Python
+    #[default]
+    Disallowed,
+    // Allow a system Python to be used if no vritual environment is active.
+    Allowed,
+    // Prefer a system Python over virtual environments.
+    Preferred,
+}
+
+/// The result of an interpreter search.
+///
+/// Returned by [`find_interpreter`].
+type InterpreterResult = Result<DiscoveredInterpreter, InterpreterNotFound>;
+#[allow(non_upper_case_globals)]
```

Comment:
I really want to separate "There was an error during discovery" from "We did not find an interpreter". They mean very different things. I could remove the aliases or change it to be a custom enum instead of a result — I explored both of these options previously though and found this the clearest.

Reply:
What are the aliases doing exactly? Am I correct that `Ok(FoundInterpreter(result))` could be replaced with `Ok(InterpreterResult::Ok(result))`? So it's just syntactic sugar?

Reply:
Yep just sugar to clarify the meaning, I'm okay with the latter (and had it earlier) but chose the current sugar because I think "Found" is clearer than "Ok". I especially don't like `Ok(Ok(...))`.

Reply:
I tried to create a more explicit `InterpreterResult::Found` alias but you can't do that since `Result` is external.

Reply:
I would personally find `Ok(Ok(...))` and `Ok(Err(...))` to be clearer, since I don't have to guess at the meaning or look back at this definition.

Reply:
Either way it's trade-offs at where you have to look when, but two votes is enough I'll just remove the aliases.

Reply:
I think it's fine in principle, I just haven't seen it before so I'm finding the unfamiliarity of it to be not worth it (but this is highly personal, others may disagree, maybe it is a common pattern outside of our crates).